### PR TITLE
CAR-2521 Use CMP0156 to handle duplicate libraries in link lines

### PR DESCRIPTION
**Pull request by bhanu-cantina**
**Issue:** When building protobuf, two linker warnings related to duplicate libraries appear toward the end of the process on macOS

**Root Cause:**
- Platform-Specific Issue: The warnings are only seen on macOS but not on Linux (verified using Jenkins workers).
- New Apple Linker: With the Xcode 15 update, Apple introduced a new linker aimed at speeding up static linking. This new linker automatically re-scans static libraries, meaning repeated static libraries in the link line are unnecessary. However, the older classic linker (-ld64) does require repeated libraries.

**Solution:**
- CMake Fix: CMake version 3.29.0 and above resolves this by de-duplicating libraries in link lines based on the linker’s capabilities. It ensures that static libraries are listed only once in the link line for platforms like macOS, where repeated libraries are no longer needed. CMake applies this through policy CMP0156.
- To enforce this, we’ve updated CMakeLists.txt to have `cmake_policy(SET CMP0156 NEW)`, ensuring that libraries are de-duplicated on supported platforms.

**Xcode 15 Release Notes:**
A new linker has been written to significantly speed up static linking. It’s the default for all macOS, iOS, tvOS, and visionOS binaries and anyone using the “Mergeable Libraries” feature. The classic linker can still be explicitly requested using -ld64, and will be removed in a future release.

**CMake 3.29.0 Release Notes:**
CMake learned to de-duplicate libraries on link lines based on linker capabilities. See policy CMP0156.
This policy was introduced in CMake version 3.29. It may be set by cmake_policy() or cmake_minimum_required().

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,10 @@ if(POLICY CMP0141)
   cmake_policy(SET CMP0141 OLD)
 endif()
 
+# airtime: start - CAR-2521 Use CMP0156 to handle duplicate libraries in link lines
+cmake_policy(SET CMP0156 NEW)
```
Comment (sumitkandpal) [inline CMakeLists.txt:11]: what this policy does?. could you  also put a note in the pr description regarding how it helps

**Comment by jasonshweiri** [comment]
Looks like protobuf is using its own abseil cpp in `third_party`. Seems like we should change it to use ours.
We can add this to aircore cmakelists:
```
--- a/aircore/CMakeLists.txt
+++ b/aircore/CMakeLists.txt
@@ -3,6 +3,34 @@ cmake_policy(SET CMP0079 NEW)
 cmake_policy(SET CMP0077 NEW)
 
 set(protobuf_ABSL_PROVIDER "none" CACHE STRING "")
+set(protobuf_ABSL_USED_TARGETS
+absl::absl_check
+absl::absl_log
+absl::algorithm
+absl::bind_front
+absl::bits
+absl::btree
+absl::cleanup
+absl::core_headers
+absl::debugging
+absl::die_if_null
+absl::dynamic_annotations
+absl::flags
+absl::flat_hash_map
+absl::flat_hash_set
+absl::function_ref
+absl::layout
+absl::log_initialize
+absl::memory
+absl::node_hash_map
+absl::node_hash_set
+absl::optional
+absl::span
+absl::statusor
+absl::type_traits
+absl::utility
+absl::variant
+)
 add_subdirectory_once(${at_media_deps_dir}/abseil-cpp ${CMAKE_BINARY_DIR}/abseil-cpp)
 set(protobuf_INSTALL NO)
```
and comment out this line from their CMakeLists:
```
# include(${protobuf_SOURCE_DIR}/cmake/abseil-cpp.cmake)
```
---
Otherwise (if we want to keep using `third_party` abseil) you can adjust `abseil-cpp.cmake` to just remove the duplicate links:
```
diff --git a/cmake/abseil-cpp.cmake b/cmake/abseil-cpp.cmake
index b50fb89e6..9013c6734 100644
--- a/cmake/abseil-cpp.cmake
+++ b/cmake/abseil-cpp.cmake
@@ -57,12 +57,10 @@ else()
     absl::absl_check
     absl::absl_log
     absl::algorithm
-    absl::base
     absl::bind_front
     absl::bits
     absl::btree
     absl::cleanup
-    absl::cord
     absl::core_headers
     absl::debugging
     absl::die_if_null
@@ -71,20 +69,14 @@ else()
     absl::flat_hash_map
     absl::flat_hash_set
     absl::function_ref
-    absl::hash
     absl::layout
     absl::log_initialize
-    absl::log_severity
     absl::memory
     absl::node_hash_map
     absl::node_hash_set
     absl::optional
     absl::span
-    absl::status
     absl::statusor
-    absl::strings
-    absl::synchronization
-    absl::time
     absl::type_traits
     absl::utility
     absl::variant
```
Thoughts?

Edit: I think I may be wrong about it using abseil from `third_party`. To fix the duplicate links either of the solutions outlined above still works though.

**Comment by bhanu-cantina** [comment]
@jasonshweiri 

- I don't think libs mentioned in the `protobuf_ABSL_USED_TARGETS` are he cause of the duplicate libs linker warnings.
- On Linux, I’m not seeing any duplicate library warnings. After tracing libabsl_time.a, I noticed several other libraries depend on it, and it’s listed as a dependency target in CMake.
- It seems to me that CMake generates link lines as usual for standard linkers, where dependencies must appear to the right side of the link line (since these linkers typically perform a single pass).
- However, the new linker introduced in Xcode 15 re-scans the link line for missing dependencies if they aren’t found on the right.

> Xcode 15 Release Notes:
> A new linker has been written to significantly speed up static linking. It's the default for all macOS, iOS, tvOS, and visionOS binaries and anyone using the "Mergeable Libraries" feature. The classic linker can still be explicitly requested using -ld64, and will be removed in a future release.
> 
> CMake 3.29.0 Release Notes:
> CMake learned to de-duplicate libraries on link lines based on linker capabilities. See policy CMP0156.
> This policy was introduced in CMake version 3.29. It may be set by cmake_policy() or cmake_minimum_required().

**Comment by jasonshweiri** [comment]
@bhanu-cantina I think I understand. Your original fix is fine with me.

File: CMakeLists.txt (hunk 1 above)
Comment (bhanu-cantina) [reply CMakeLists.txt:11]: - With `CMP0156` policy set to new, cmake will generate link lines based on the linker capabilities. Updated PR description.

**Review by jasonshweiri — APPROVED** [review]

**Pull request merged**